File: leetcodejava/src/main/java/com/leetcode/OneSixTwo.java
package com.leetcode;

/**
 * @author: John
 * @date: 2020-05-11 11:56
 * @description: 162
 **/
public class OneSixTwo {


    public int findPeakElement(int[] nums) {
        int l = 0, r = nums.length - 1;
        while (l < r) {
            int mid = l + (r - l) / 2;
            if (nums[mid] > nums[mid + 1]) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }
        return l;
    }

    /**
     * recursive binary search
     *
     * @param nums
     * @return
     */
//    public int findPeakElement(int[] nums) {
//        return search(nums, 0, nums.length - 1);
//    }
//
//    public int search(int[] nums, int l, int r) {
//        if (l == r) {
//            return l;
//        }
//        int mid = l + (r - l) / 2;
//        if (nums[mid] > nums[mid + 1]) {
//            return search(nums, l, mid);
//        }
//        return search(nums, mid + 1, r);
//    }

    // linear scan
//    public int findPeakElement(int[] nums) {
//        for (int i = 0; i < nums.length - 1; i++) {
//            if (nums[i] > nums[i + 1]) {
//                return i;
//            }
//        }
//        return nums.length - 1;
//    }


}

File: leetcodecpp/leetcode/OneSixTwo.cpp
//
// Created by John on 2020/2/29.
//
#include "vector"

using namespace std;

class Solution{

public:
    // brute solution linear scan
//    int findPeakElement(vector<int>& nums) {
//        for (int i = 0; i < nums.size() - 1; i++) {
//            if (nums[i] > nums[i + 1]) {
//                return i;
//            }
//        }
//        return nums.size() - 1;
//    }

    // recursion binary search
//    int findPeakElement(vector<int> & nums) {
//        return search(nums, 0, nums.size() - 1);
//    }
//    int search(vector<int> & nums, int l, int r) {
//        if (l == r) {
//            return l;
//        }
//        int mid = l + (r - l) / 2;
//        if (nums[mid] > nums[mid + 1]) {
//            return search(nums, l, mid);
//        }
//        return search(nums, mid + 1, r);
//    }

    // iterative binary search
    int findPeakElement(vector<int> & nums) {
        int l = 0, r = nums.size() - 1;
        while (l < r) {
            int mid = l + (r - l) / 2;
            if (nums[mid] > nums[mid + 1]) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }
        return l;
    }

};

//int main() {
//    vector<int> nums;
//    nums.push_back(1);
//    nums.push_back(2);
//    nums.push_back(1);
//    nums.push_back(3);
//    nums.push_back(5);
//    nums.push_back(6);
//    nums.push_back(4);
//    Solution * solution = new Solution();
//    solution->findPeakElement(nums);
//}